tests/expectations/test_dataclass_serializable_dot_dict_pattern.py

```diff
@@ -0,0 +1,507 @@
+                    "quux": 43,
+                }
+            )
+        ],
+        enum_list=[
+            MyEnum("x"),
+            MyEnum("y"),
+            MyEnum("z"),
+        ]
+    )
+
+    C_dict = my_C.to_dict()
+
+    # Make sure it's a dictionary, not a DictDot
+    assert type(C_dict) == dict
+    assert isinstance(C_dict, DictDot) == False
+    # Dictionaries don't support dot notation.
+    with raises(AttributeError):
+        C_dict.A_list
+
+    assert type(C_dict["A_list"][0]) == dict
+    assert type(C_dict["B_list"][0]) == dict
+
+    assert C_dict == {
+        "alpha_var": 20,
+        "beta_var": "x",
+        "A_list": [
+            {
+                "foo": "A-1",
+                "bar": 101,
+            },
+            {
+                "foo": "A-2",
+                "bar": 102,
+            },
+        ],
+        "B_list": [
+            {
+                "foo": "B-1",
+                "bar": 201,
+                "baz": ["a", "b", "c"],
+                "qux": -100,
+                "quux": 43,
+            }
+        ],
+        "enum_list": [
+            "x",
+            "y",
+            "z",
+        ]
+    }
+
+
+def test_immutability():
+    "Can be made immutable"
+
+    @dataclass(frozen=True)
+    class MyClassD(SerializableDictDot):
+        foo: str
+        bar: int
+
+    my_D = MyClassD(
+        **{
+            "foo": "a string",
+            "bar": 1,
+        }
+    )
+    assert my_D["foo"] == "a string"
+    assert my_D["bar"] == 1
+
+    with raises(FrozenInstanceError):
+        my_D.foo = "different string"
+
+    assert my_D["foo"] == "a string"
+    assert my_D.foo == "a string"
+
+    with raises(FrozenInstanceError):
+        my_D["foo"] = "a third string"
+
+    assert my_D["foo"] == "a string"
+    assert my_D.foo == "a string"
+
+
+def test_reserved_word_key():
+    """Can be instantiated with a key that's also a reserved word
+    
+This pattern is helpful for cases where we're migrating from dictionary-based objects to typed objects,
+and the dictionary contains keys that are reserved words in python.
+
+For example, test cases use the reserved word: "in" as one of their required fields.
+    """
+
+    @dataclass
+    class MyClassE(SerializableDictDot):
+        foo: str
+        bar: int
+        input: int
+
+    class MyClassF(MyClassE):
+        def __init__(
+            self,
+            foo,
+            bar,
+            **kwargs,
+        ):
+            super().__init__(foo=foo, bar=bar, input=kwargs["in"])
+
+    my_F = MyClassF(
+        foo="a string",
+        bar=1,
+        **{"in": 10},
+    )
+    assert my_F["foo"] == "a string"
+    assert my_F["bar"] == 1
+    assert my_F["input"] == 10
+    assert my_F.input == 10
+
+    my_F = MyClassF(
+        **{
+            "foo": "a string",
+            "bar": 1,
+            "in": 10,
+        }
+    )
+    assert my_F["foo"] == "a string"
+    assert my_F["bar"] == 1
+    assert my_F["input"] == 10
```

Comment:
add: assert my_F["in"] == 10

Reply:
I'm worried we lose symmetry here: I gave you a dictionary with one set of keys, you're giving me back something different. How would the user know to access using ["input"] when they provided ["in"]?